File: system-test/dotnet/SystemTests/Latest/ArchitectureTests/ArchitectureRulesTest.cs
using ArchUnitNET.Domain;
using ArchUnitNET.Domain.Extensions;
using ArchUnitNET.Fluent;
using ArchUnitNET.Fluent.Conditions;
using ArchUnitNET.Loader;
using ArchUnitNET.xUnit;
using Xunit;
using static ArchUnitNET.Fluent.ArchRuleDefinition;

namespace SystemTests.Latest.ArchitectureTests;

/// <summary>
/// POC parity with Java's ArchitectureRulesTest: which DSL/driver structural rules can be
/// enforced mechanically? Four representative rules (A1/A2/A7/A10) via ArchUnitNET.
/// </summary>
public class ArchitectureRulesTest
{
    private static readonly Architecture TestKit = new ArchLoader()
        .LoadAssemblies(
            typeof(Driver.Port.IMyShopDriver).Assembly,
            typeof(Dsl.Core.Shared.VoidVerification).Assembly,
            typeof(Dsl.Core.UseCase.UseCases.ViewOrderVerification).Assembly)
        .Build();

    // A1 — request DTOs in the driver port expose only string members.
    [Fact]
    [Trait("Category", "Architecture")]
    public void RequestDtosDeclareOnlyStringMembers()
    {
        IArchRule rule = Classes().That()
            .ResideInNamespace("Driver.Port.Dtos").And()
            .HaveNameEndingWith("Request")
            .Should().FollowCustomCondition(cls =>
            {
                var nonString = cls.GetPropertyMembers()
                    .Where(p => p.Type.Name != "String")
                    .ToList();
                var pass = nonString.Count == 0;
                var failure = pass
                    ? ""
                    : $"{cls.FullName} declares non-string members: "
                      + string.Join(", ", nonString.Select(p => $"{p.Name}:{p.Type.Name}"));
                return new ConditionResult(cls, pass, failure);
            }, "Request DTOs in Driver.Port.Dtos must declare only string members");

        rule.Check(TestKit);
    }

    // A2 — verification public methods are fluent (own type) or terminal (void); never getters.
    [Fact]
    [Trait("Category", "Architecture")]
    public void VerificationMethodsAreFluentOrVoid()
    {
        IArchRule rule = Classes().That()
            .ResideInNamespaceMatching("Dsl\\.Core").And()
            .HaveNameEndingWith("Verification")
            .Should().FollowCustomCondition(cls =>
            {
                var violations = cls.GetMethodMembers()
                    .Where(m => m.Visibility == Visibility.Public)
                    .Where(m => m.ReturnType.FullName != cls.FullName && m.ReturnType.Name != "Void")
                    .ToList();
                var pass = violations.Count == 0;
                var failure = pass
                    ? ""
                    : $"{cls.FullName} has non-fluent public methods: "
                      + string.Join(", ", violations.Select(m => $"{m.Name}->{m.ReturnType.Name}"));
                return new ConditionResult(cls, pass, failure);
            }, "Verification public methods must return their own type (fluent) or void (terminal) — no getters");

        rule.Check(TestKit);
    }

    // A7 — DSL core declares no own *Request/*Response; it reuses Driver.Port.Dtos.
    [Fact]
    [Trait("Category", "Architecture")]
    public void DslCoreDeclaresNoOwnRequestOrResponseDtos()
    {
        IArchRule rule = Classes().That()
            .ResideInNamespaceMatching("Dsl\\.Core")
            .Should().NotHaveNameEndingWith("Request")
            .AndShould().NotHaveNameEndingWith("Response");

        rule.Check(TestKit);
    }

    // A10 — every IMyShopDriver method takes a single *Request and returns Result<*Response, …> (strict).
    [Fact]
    [Trait("Category", "Architecture")]
    public void EveryDriverMethodTakesRequestAndReturnsResponse()
    {
        IArchRule rule = Interfaces().That()
            .HaveName("IMyShopDriver")
            .Should().FollowCustomCondition(driver =>
            {
                var problems = new List<string>();
                foreach (var method in driver.GetMethodMembers())
                {
                    var parameters = method.Parameters.ToList();
                    if (parameters.Count != 1 || !parameters[0].Name.EndsWith("Request"))
                    {
                        problems.Add(
                            $"{method.Name} must take exactly one *Request parameter "
                            + $"(was [{string.Join(", ", parameters.Select(p => p.Name))}])");
                    }

                    var returnsResponse = method.Dependencies.Any(d => d.Target.Name.EndsWith("Response"));
                    if (!returnsResponse)
                    {
                        problems.Add($"{method.Name} must return Result<*Response, …>");
                    }
                }

                var pass = problems.Count == 0;
                return new ConditionResult(driver, pass, string.Join("; ", problems));
            }, "Every IMyShopDriver method must take a single *Request and return Result<*Response, …>");

        rule.Check(TestKit);
    }
}

File: system-test/java/src/test/java/com/mycompany/myshop/systemtest/architecture/ArchitectureRulesTest.java
package com.mycompany.myshop.systemtest.architecture;

import com.tngtech.archunit.core.domain.JavaClasses;
import com.tngtech.archunit.core.domain.JavaMethod;
import com.tngtech.archunit.core.domain.JavaParameterizedType;
import com.tngtech.archunit.core.domain.JavaType;
import com.tngtech.archunit.core.importer.ClassFileImporter;
import com.tngtech.archunit.core.importer.ImportOption;
import com.tngtech.archunit.lang.ArchCondition;
import com.tngtech.archunit.lang.ArchRule;
import com.tngtech.archunit.lang.ConditionEvents;
import com.tngtech.archunit.lang.SimpleConditionEvent;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

import static com.tngtech.archunit.core.domain.JavaClass.Predicates.resideInAPackage;
import static com.tngtech.archunit.core.domain.JavaClass.Predicates.simpleNameEndingWith;
import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.fields;
import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.methods;
import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.noClasses;

/**
 * POC: which DSL/driver structural rules can ArchUnit enforce mechanically?
 * Four representative rules, one per feasibility tier. See plan
 * 20260620-0741-archunit-enforce-dsl-driver-rules-investigation.md.
 */
@Tag("architecture")
class ArchitectureRulesTest {

    private static final JavaClasses TESTKIT = new ClassFileImporter()
            .withImportOption(ImportOption.Predefined.DO_NOT_INCLUDE_TESTS)
            .importPackages("com.mycompany.myshop.testkit");

    /** A1 — request DTOs in the driver port expose only String fields. */
    @Test
    void requestDtosDeclareOnlyStringFields() {
        ArchRule rule = fields()
                .that().areDeclaredInClassesThat(
                        resideInAPackage("..driver.port.dtos..").and(simpleNameEndingWith("Request")))
                .and().areNotStatic()
                .should().haveRawType(String.class)
                .as("Request DTOs in driver.port.dtos must declare only String fields");

        rule.check(TESTKIT);
    }

    /** A2 — verification public methods are fluent (own type) or terminal (void); never getters. */
    @Test
    void verificationMethodsAreFluentOrVoid() {
        ArchRule rule = methods()
                .that().areDeclaredInClassesThat(
                        resideInAPackage("..dsl.core..").and(simpleNameEndingWith("Verification")))
                .and().arePublic()
                .should(returnOwningClassOrVoid())
                .as("Verification public methods must return their own type (fluent) or void (terminal) — no getters");

        rule.check(TESTKIT);
    }

    /** A7 — DSL core declares no own *Request/*Response; it reuses driver.port.dtos ("identical req/resp"). */
    @Test
    void dslCoreDeclaresNoOwnRequestOrResponseDtos() {
        ArchRule rule = noClasses()
                .that().resideInAPackage("..dsl.core..")
                .should().haveSimpleNameEndingWith("Request")
                .orShould().haveSimpleNameEndingWith("Response")
                .as("DSL core must not declare its own *Request/*Response DTOs — it shares driver.port.dtos");

        rule.check(TESTKIT);
    }

    /** A10 — every MyShopDriver operation takes a *Request and returns Result<*Response, …> (strict, Q2(b)). */
    @Test
    void everyDriverMethodTakesRequestAndReturnsResponse() {
        ArchRule rule = methods()
                .that().areDeclaredInClassesThat().haveSimpleName("MyShopDriver")
                .should(takeRequestAndReturnResponse())
                .as("Every MyShopDriver method must take a single *Request and return Result<*Response, …>");

        rule.check(TESTKIT);
    }

    private static ArchCondition<JavaMethod> returnOwningClassOrVoid() {
        return new ArchCondition<>("return their declaring type (fluent) or void (terminal)") {
            @Override
            public void check(JavaMethod method, ConditionEvents events) {
                String returnType = method.getRawReturnType().getName();
                String owner = method.getOwner().getName();
                boolean ok = returnType.equals("void") || returnType.equals(owner);
                if (!ok) {
                    events.add(SimpleConditionEvent.violated(method,
                            method.getFullName() + " returns " + returnType + " (expected " + owner + " or void)"));
                }
            }
        };
    }

    private static ArchCondition<JavaMethod> takeRequestAndReturnResponse() {
        return new ArchCondition<>("take a single *Request and return Result<*Response, …>") {
            @Override
            public void check(JavaMethod method, ConditionEvents events) {
                var params = method.getRawParameterTypes();
                if (params.size() != 1 || !params.get(0).getSimpleName().endsWith("Request")) {
                    events.add(SimpleConditionEvent.violated(method,
                            method.getFullName() + " must take exactly one *Request parameter (was " + params + ")"));
                }

                boolean responseOk = false;
                JavaType returnType = method.getReturnType();
                if (returnType instanceof JavaParameterizedType parameterized
                        && !parameterized.getActualTypeArguments().isEmpty()) {
                    String firstArg = parameterized.getActualTypeArguments().get(0).toErasure().getSimpleName();
                    responseOk = firstArg.endsWith("Response");
                }
                if (!responseOk) {
                    events.add(SimpleConditionEvent.violated(method,
                            method.getFullName() + " must return Result<*Response, …> (was " + returnType.getName() + ")"));
                }
            }
        };
    }
}
